Add convert_bufr suite (ioda-converters)

cohen-seth opened the pull request:
## Description
Initial version of the new get_bufr and bufr_to_ioda tasks with a new suite convert_bufr. 

_This suite is INCOMPLETE (intentionally): currently it can only convert BUFR files of one observation type (AMSUA)._

This suite currently relies upon Jedi Ioda-converter (bufr2ioda.x), yaml configuration files to use in the bufr2ioda conversion (configuration/jedi/iodaconv) , CloneGeosMksi task (for the .rc database file), CloneGMAOPerllib (for use_acquire.py) previous SWELL runs and stored IODA outputs in R2D2.

`get_bufr` -> Gets the BUFR data files:
- Using the paths in the GEOS_mksi/ObsClass/obsys-nccs.rc database, it identifies the path(s) of BUFR files for a given observation type (AMSUA), then creates a sub directory within the cycle directory where it stores copies of the BUFR files.

`bufr_to_ioda` -> Converts the BUFR files from get_bufr to Ioda format using the Jedi Ioda-converter `bufr2ioda.x`:
- Generates conversion yamls for the BUFR files from `get_bufr`, then using those conversion yamls it converts the files to Ioda format 

## Conversion yaml
**This yaml is used in the conversion step:** 
```sh
bufr2ioda.x [conversion yaml]
```
The converter, `bufr2ioda.x`, requires a yaml file (**conversion yaml**) containing:

1. the path of the input BUFR file to convert
2. the path where it will save the converted ioda (.nc4) file
3. the information about the variables in the BUFR file that it will look for and convert. 

One of these conversion yaml files is needed at each of the conversion steps for each file - i.e. it loops through and converts the files one by one and needs a new yaml file each time. 

`bufr_to_ioda` uses a **configuration yaml** to generate the **conversion yaml**

## Configuration yaml
**This yaml is used to GENERATE the conversion yaml**
```sh
copy the configuration yaml to cycle directory > make edits > save file. The resulting file is the conversion yaml.
```

A **configuration yaml** file to use for creating the conversion yamls has been added to configuration/jedi/iodaconv. These yamls are specific to observation type. 

Currently, the conversion yaml is overwritten each time it converts a new file.

## Known Issues/To-do:
- Clean & document code
- Need to fix the flow.cylc - specifically, the CloneGeosMksi step. I cloned it manually rather than debug in order to save time.
- Add more observation types - generalizing the code, adding configuration yamls for each additional obs type


## Dependencies
- Jedi iodaconverter ~ bufr2ioda.x
- Ioda conversion yaml ~  used in the conversion step: bufr2ioda.x [conversion yaml]
- Configuration yaml ~ A template to use for creating the conversion yaml(s). Based on Jedi iodaconv/test/testinput/[*observation type*].yaml
- GeosMksi ~ ObsClass/obsys-nccs.rc database file for paths
- UseAcquire ~ GmaoPerlib

cohen-seth committed:
Add iodaconverters for bufr files.

cohen-seth committed:
Adding feature: iodaconverters for bufr files.

cohen-seth committed:
Bufr file search config fix

cohen-seth committed:
Update branch with latest develop version

Merge branch 'develop' into feature/iodaconverters

mranst committed:
Add task: CloneGmaoPerllib. Allow use of acquire and acquire_obsys for convert_bufr suite

mranst committed:
Merge remote-tracking branch 'origin/develop' into feature/use_acquire

mranst committed:
Fixed environment for UseAcquire, pycodestyle fix

mranst committed:
Made path_to_bufr a platform-dependent question

cohen-seth committed:
Merge branch 'feature/mranst/use_acquire' into feature/iodaconverters

cohen-seth committed:
Pulling only files in daterange not entire directory.

cohen-seth committed:
Pulling only files in daterange not entire directory.

cohen-seth committed:
Merge branch 'develop' into feature/iodaconverters

Update feature/iodaconverters

cohen-seth committed:
added bufr path to test yaml

cohen-seth committed:
Merge branch 'feature/iodaconverters' of https://github.com/GEOS-ESM/swell into feature/iodaconverters

cohen-seth committed:
restoring changes to yaml file.

cohen-seth committed:
get_bufr generic filename fix.

cohen-seth committed:
get_bufr remove comments.

cohen-seth committed:
fix for cloning perllib

cohen-seth committed:
Merge branch 'develop' into feature/iodaconverters

cohen-seth committed:
get bufr use acquire method

cohen-seth committed:
merging

cohen-seth committed:
BufrToIoda working version 1.0

cohen-seth committed:
Merge branch 'develop' into feature/iodaconverters

Updating feature/iodaconverters to be up to date with develop for pull request

Dooruk commented on src/swell/suites/convert_bufr/flow.cylc:
```diff
@@ -0,0 +1,120 @@
+
+            # Convert BUFR to ioda
+            GetBufr
+            GetBufr => BufrToIoda
```
```suggestion
            GetBufr => BufrToIoda
```
This is the concise way of doing it

mranst replied:
I think you want `CloneGeosMksi` instead of/in addition to `CloneGeos` to get the .rc file right?

cohen-seth replied:
@mranst You are right. I was having trouble doing it with the flow.cylc file. Mentioned in the PR description issues. I just made the corrections of 'CloneGeosMksi' rather than 'CloneGeos' in the description too.

Dooruk commented on src/swell/suites/convert_bufr/flow.cylc:
```diff
@@ -0,0 +1,120 @@
+
+    # GetYaml ~ existing code outside of Swell workflow
+    # EditYaml
+
```
```suggestion
```

Dooruk commented on src/swell/suites/convert_bufr/flow.cylc:
```diff
@@ -0,0 +1,120 @@
+            --{{key}} = {{value}}
+        {%- endfor %}
+
+    [[ GetBufr ]]
```
```suggestion
    [[GetBufr]]
```
Nitpick but can you get rid of these spaces

Dooruk commented on src/swell/suites/convert_bufr/flow.cylc:
```diff
@@ -0,0 +1,120 @@
+    # GetYaml ~ existing code outside of Swell workflow
+    # EditYaml
+
+    [[ BufrToIoda ]]
```
```suggestion
    [[BufrToIoda]]
```

Dooruk commented on src/swell/suites/suite_questions.yaml:
This file is irrelevant after Michael's recent updates and needs to be erased.

Dooruk commented on src/swell/tasks/bufr_to_ioda.py:
```diff
@@ -0,0 +1,287 @@
+
+
+import os
+import subprocess
```
There is a `subprocess` wrapper utility with different subprocess methods available, maybe not now but on the next iteration you can use that:

https://github.com/GEOS-ESM/swell/blob/develop/src/swell/utilities/shell_commands.py

Dooruk commented on src/swell/tasks/bufr_to_ioda.py:
```diff
@@ -0,0 +1,287 @@
+#path_to_ioda_conv_yaml_tmpl_dir = os.path.join(self.experiment_path(), 'configuration/iodaconv')
+
+#bufr_to_ioda_yaml_obs_type_dict = {
+bufr2ioda_obs_type_dict = {
```
This dictionary logic is good. Are these keys you are using (atms, 1bamua) from GSI nomenclature? I could use @rtodling's input here too.

rtodling replied:
From what I can tell these keys are consistent w/ the GEOS-FP (atmospheric DAS) observations database.

cohen-seth replied:
@Dooruk They are from the/our naming convention the Bufr files - I'm not sure if it's the GSI nomenclature. They are how the respective obs types are represented in the filenames. 

Ex:

1. gdas1.180101.t00z.gpsro.tm00.bufr_d
2. gdas1.20200101.t00z.atms.tm00.bufr_d
3. gdas1.%Y%m%d.t%Hz.[obs type].tm00.bufr_d

(obs type = key)

Dooruk commented on src/swell/tasks/bufr_to_ioda.py:
```diff
@@ -0,0 +1,287 @@
+# --------------------------------------------------------------------------------------------------
+
+# Dictionary linking each obs type to the appropriate yaml template
+#path_to_ioda_conv_yaml_tmpl_dir = os.path.join(self.experiment_path(), 'configuration/iodaconv')
```
I understand this is in draft format but could you clear extra comments? Makes it hard to follow the code 😄

Dooruk commented on src/swell/tasks/bufr_to_ioda.py:
```diff
@@ -0,0 +1,287 @@
+            print(f"Match found: {part}")
+            return part
+    print("No match found.")
+    return None
```
Why not use the dict directly?

```suggestion
    parts = filename.split('.')
    for part in parts:
        if part in bufr2ioda_obs_type_dict:
            print(f"Match found: {part}")
            return part
    print("No match found.")
    return None
```

Also try using `self.logger.info` instead of `print` for consistency.

Dooruk commented on src/swell/tasks/get_bufr.py:
```diff
@@ -0,0 +1,132 @@
+        os.makedirs(bufr_dir, 0o755, exist_ok=True)
+
+
+        ihms = '060000' # window_length in experiment.yaml
```
While this won't change for atmosphere, still try to create this by using `window_length` from your qd.

cohen-seth replied:
@Dooruk For clarification, you're referring to `ihms` here correct?

Dooruk commented on src/swell/tasks/get_bufr.py:
```diff
@@ -0,0 +1,132 @@
+
+
+import os
+import subprocess
```
same comment about `subprocess` wrapper

cohen-seth replied:
@Dooruk Will need some help applying existing utils. I think we discussed this previously.

Dooruk commented on src/swell/tasks/get_bufr.py:
```diff
@@ -0,0 +1,132 @@
+        geos_mksi_obsysrc_path = os.path.join(swell_exp_path, 'GEOS_mksi/ObsClass/obsys-nccs.rc')
+
+        # 
+        env_dict = os.environ
+        env_dict['PATH'] = env_dict["PATH"] + ":" + perl_executable_path
```
Cylc allows task specific environment, it is a bit involved approach and requires understanding Task configuration but it is explained a bit here:

https://cylc.github.io/cylc-doc/stable/html/user-guide/writing-workflows/runtime.html#workflow-share-directories

cohen-seth replied:
@Dooruk I'm not sure I know what you mean or what it is that needs to be changed.

Dooruk commented on src/swell/tasks/get_bufr_iodaconv_yaml.py:
Isn't this part of `get_bufr` now?

Dooruk commented on src/swell/tasks/use_acquire.py:
Also extra here?

Dooruk commented on src/swell/test/suite_tests/convert_bufr-tier1.yaml:
This file is obsolete

Dooruk commented on .gitignore:
```diff
@@ -137,3 +137,7 @@ dmypy.json
+# sicohen dev directories and scrap work
+sic_scrap/
+src/swell/Swell_sles15/src/swell/tasks/get_bufr_no_acuqire.py
+src/swell/Swell_sles15/src/swell/tasks/get_bufr_iodaconv_yaml.py
```
Hmm

Dooruk reviewed (changes requested):
@cohen-seth great job getting this complicated PR in.

As you noted, this requires some cleaning of redundant bits, mostly having to do with @mranst's changes with task/suite yamls to task/suite dictionaries.

The key part is mapping observations to bufr names, I need to talk to @rtodling about this as I'm not familiar with the GSI world. In the meantime I added some comments for short term changes.

mranst commented on src/swell/tasks/clone_gmao_perllib.py:
```diff
@@ -0,0 +1,82 @@
+                             + "nodes with no internet access.")
+            subprocess.run(f"git clone -b {gmao_perllib_tag} "
+                           + "https://github.com/GEOS-ESM/GMAO_perllib.git "
+                           + os.path.join(self.experiment_path(),"GMAO_perllib"))
```
```suggestion
            subprocess.run(['git', 'clone', '-b', gmao_perllib_tag,
                            'https://github.com/GEOS-ESM/GMAO_perllib.git',
                             os.path.join(self.experiment_path(), 'GMAO_perllib')])
```

mranst replied:
I'm looking back at my additions and realizing some things need to be fixed

mranst commented on src/swell/tasks/task_questions.py:
I think there's a lot of questions that are invoked here that aren't referenced in your code. Only the ones that are invoked using `self.config.<field>` need to be added to this list

cohen-seth replied:
@mranst You're right. Had some trouble with these. Made additions trying to get the flow.cylc to set everything up properly but couldn't; then did not delete them.

mranst commented on src/swell/tasks/task_questions.py:
```diff
@@ -642,3 +642,37 @@ class TaskQuestions(QuestionContainer, Enum):
+    )
+    
+    # --------------------------------------------------------------------------------------------------
+
```
I can't suggest the changes in the appropriate place, but `CloneGMAOPerllib` needs these additions:

`utilities/question_defaults.py`:
```python
class existing_perllib_path(TaskQuestion):
    default_value: str = 'defer_to_platform'
    question_name: str = 'existing_perllib_path'
    models: List[str] = mutable_field([
        'geos_atmosphere'
    ])
    prompt: str = "Provide a path to an existing location for GMAO_perllib."
    widget_type: WType = WType.STRING

class gmao_perllib_tag(TaskQuestion):
    default_value: str = 'g1.0.1'
    question_name: str = gmao_perllib_tag
    models: List[str] = mutable_field([
        'geos_atmosphere'
    ])
    prompt: str = "Specify the tag at which GMAO_perllib should be cloned."
    widget_type: WType = WType.STRING
```

`task_questions.py`:
```python
CloneGMAOPerllib = QuestionList(
    list_name="CloneGMAOPerllib",
    questions=[
        qd.existing_perllib_path(),
        qd.gmao_perllib_tag()
    ]
)
```

cohen-seth replied:
@mranst We'll need to discuss. I believe these may be address by fixing the flow.cylc.

mranst reviewed (dismissed):
Nice work, this is quite a complex task to tackle!

rtodling commented:
I just quickly looked at it ... I want to have a chance pull this branch on my end; and try it out to get true feel for how this is working. I will be on leave for most of next week; though I might have a changes to work on it one day next week.

cohen-seth committed:
Update src/swell/suites/convert_bufr/flow.cylc

Co-authored-by: Doruk Ardağ <38666458+Dooruk@users.noreply.github.com>

cohen-seth committed:
Update src/swell/suites/convert_bufr/flow.cylc

Co-authored-by: Doruk Ardağ <38666458+Dooruk@users.noreply.github.com>

cohen-seth committed:
Update .gitignore

cohen-seth committed:
Update bufr_to_ioda.py

Cleaning up comments left from testing

cohen-seth committed:
Update bufr_to_ioda.py

more clean up

cohen-seth committed:
Delete src/swell/suites/suite_questions.yaml

cohen-seth committed:
Apply suggestions from code review

Co-authored-by: Doruk Ardağ <38666458+Dooruk@users.noreply.github.com>

cohen-seth committed:
Apply suggestions from code review

Co-authored-by: Michael Anstett <michael.anstett@nasa.gov>

cohen-seth committed:
Update get_bufr.py

cohen-seth reviewed (commented):
Addressed suggestions and corrections of the initial review of the PR.

Dooruk committed:
few fixes for bufr converter (#597)

* Update JEDI version to July 4, 2025 (#593)

* Swell doc updates for offline installation of Swell and uv (#588)

* add instructions for offline installation of Swell and Uv

* fix the bug of missed links

* Add explanation based on Alex's suggestion.

* fix a link otherwise causes 404 error

* revise steps SWELL installtion with more details

* fix code error in soca 3dfgat_cycle, and incorrect
link that leads to 404 error

* update steps for offline installation

* Update offline installation manual for csh users

---------

Co-authored-by: Doruk Ardağ <38666458+Dooruk@users.noreply.github.com>

* Remove delp from bkg; avoid confusion (#578)

Co-authored-by: Doruk Ardağ <38666458+Dooruk@users.noreply.github.com>

* Fix for non-captured ci test output (#596)

* add default obs bufr classes

* make fixes to the suite files

* use obs bufr classes

* fixes for two suite tasks

* erase unused file

* erase another unused file

* Clone Perllib related entries

* create proper dependencies for bufr to function as intended

---------

Co-authored-by: Michael Anstett <michael.anstett@nasa.gov>
Co-authored-by: Cheng Da <111095732+gmao-cda@users.noreply.github.com>
Co-authored-by: rtodling <ricardo.todling@nasa.gov>

Dooruk commented:
@cohen-seth is this branch ready for a test run by @rtodling in it's current form?

cohen-seth committed:
bufr_to_ioda.py fixes for cylc. Added iodaconv yamls for more obs types.

rtodling reviewed (commented):
I think the directory named iodaconv needs to be renamed to bufr2ioda.

rtodling reviewed (commented):
This is definitely working nicely.

rtodling commented:
Just a general comment. This is what I see in the *yaml files inside iodaconv:

-rwxr-xr-x 1 rtodling advda 5140 Sep 12 12:12 bufr_ncep_1bamua_n15.yaml
-rwxr-xr-x 1 rtodling advda 5317 Sep 12 12:12 bufr_ncep_1bamua_ta.yaml
-rwxr-xr-x 1 rtodling advda 4729 Sep 12 12:12 bufr_ncep_1bmhs.yaml
-rwxr-xr-x 1 rtodling advda 4603 Sep 12 12:12 bufr_ncep_atms.yaml
-rwxr-xr-x 1 rtodling advda 5098 Sep 12 12:12 bufr_ncep_atms_remap.yaml
-rwxr-xr-x 1 rtodling advda 5151 Sep 12 12:12 bufr_ncep_atms_ta_remap.yaml
-rwxr-xr-x 1 rtodling advda 6770 Sep 12 12:12 bufr_ncep_mtiasi.yaml
-rw-r--r-- 1 rtodling advda 9448 Sep 12 12:12 bufr_ncep_prepbufr_aircft.yaml
-rwxr-xr-x 1 rtodling advda 4796 Sep 12 12:12 bufr_ncep_satwind_avhrr.yaml
-rw-r--r-- 1 rtodling advda 3449 Sep 12 12:12 bufr_ncep_sevcsr.yaml


Notice that expect for aircft and severi all other have x-permission. Please remove x-permission on the files.

chmod gou-x *yaml

and they'll need to be committed.

mranst committed:
few updates to iodaconverters (#627)

* Update JEDI version to July 4, 2025 (#593)

* Swell doc updates for offline installation of Swell and uv (#588)

* add instructions for offline installation of Swell and Uv

* fix the bug of missed links

* Add explanation based on Alex's suggestion.

* fix a link otherwise causes 404 error

* revise steps SWELL installtion with more details

* fix code error in soca 3dfgat_cycle, and incorrect
link that leads to 404 error

* update steps for offline installation

* Update offline installation manual for csh users

---------

Co-authored-by: Doruk Ardağ <38666458+Dooruk@users.noreply.github.com>

* Remove delp from bkg; avoid confusion (#578)

Co-authored-by: Doruk Ardağ <38666458+Dooruk@users.noreply.github.com>

* Fix for non-captured ci test output (#596)

* Link akbk*.nc4 to the fv3files/ directory (#602)

* SWELL now controls on peripheral files (#603)

* Add task to get observations from GEOS location (#577)

* Initial commit

* intermediate

* rename task and add toggle

* restore get_observations

* fix toggle

* Updated to link

* Update naming and add task to 3dvar_atmos

* add task

* Add task to other suites

* add runtime for getobsnotinr2d2

* Automatically set window length for cycling (#604)

* Automatically set window length

* Update 3dfgat_cycle, remove comment

* Adding cloud seeding option to all-sky radiance yaml files (#610)

* add cloud seeding option to all-sky radiance yaml files

* Add cloud seeding option to MHS yaml files

---------

Co-authored-by: gmao-yzhu <gmao-yzhu@users.noreply.github.com>

* (Bugfix) Check for string None in stage_jedi (#613)

* Check for string None

* fix other tasks

* Using EVA timeseries (#549)

* works, backup before merging develop

* new eva suite

* fix getncdiags task

* erase redundant files

* add task related questions and reorder the list of task dictionaries

* new tasks, new entries

* eva capabilities configs

* comment out channel related entries for now

* erase old file

* add atmosphere YAML and have model-specific option for the suite

* atmosphere channels

* atmosphere works

* some fixes, ignore certain atmosphere outputs

* add timestamp to cycle_done

* minor changes

* erase line

* minor fix

* fix the call

* gsi nlats

* gsibec coordinates

* Test fixes for iodaconverters

* Delete use_acquire

* Refactor bufr_to_ioda

* add jinja templating to config files

* code test fixes

---------

Co-authored-by: Cheng Da <111095732+gmao-cda@users.noreply.github.com>
Co-authored-by: Doruk Ardağ <38666458+Dooruk@users.noreply.github.com>
Co-authored-by: rtodling <ricardo.todling@nasa.gov>
Co-authored-by: Yonggang Yu <yonggang.g.yu@gmail.com>
Co-authored-by: gmao-yzhu <Yanqiu.Zhu@nasa.gov>
Co-authored-by: gmao-yzhu <gmao-yzhu@users.noreply.github.com>

Dooruk commented:
@cohen-seth, sounds like from the discussion yesterday this is ready to go in so let's move this forward. There are a few steps left, the conflicts must be resolved, coding norms must pass and Swell code tests should work. Swell documentation has information about the last two, let me or @mranst know if you need help with the first item.

mranst committed:
merge develop into feature/iodaconverters (#634)

* Update JEDI version to July 4, 2025 (#593)

* Swell doc updates for offline installation of Swell and uv (#588)

* add instructions for offline installation of Swell and Uv

* fix the bug of missed links

* Add explanation based on Alex's suggestion.

* fix a link otherwise causes 404 error

* revise steps SWELL installtion with more details

* fix code error in soca 3dfgat_cycle, and incorrect
link that leads to 404 error

* update steps for offline installation

* Update offline installation manual for csh users

---------

Co-authored-by: Doruk Ardağ <38666458+Dooruk@users.noreply.github.com>

* Remove delp from bkg; avoid confusion (#578)

Co-authored-by: Doruk Ardağ <38666458+Dooruk@users.noreply.github.com>

* Fix for non-captured ci test output (#596)

* Link akbk*.nc4 to the fv3files/ directory (#602)

* SWELL now controls on peripheral files (#603)

* Add task to get observations from GEOS location (#577)

* Initial commit

* intermediate

* rename task and add toggle

* restore get_observations

* fix toggle

* Updated to link

* Update naming and add task to 3dvar_atmos

* add task

* Add task to other suites

* add runtime for getobsnotinr2d2

* Automatically set window length for cycling (#604)

* Automatically set window length

* Update 3dfgat_cycle, remove comment

* Adding cloud seeding option to all-sky radiance yaml files (#610)

* add cloud seeding option to all-sky radiance yaml files

* Add cloud seeding option to MHS yaml files

---------

Co-authored-by: gmao-yzhu <gmao-yzhu@users.noreply.github.com>

* (Bugfix) Check for string None in stage_jedi (#613)

* Check for string None

* fix other tasks

* Using EVA timeseries (#549)

* works, backup before merging develop

* new eva suite

* fix getncdiags task

* erase redundant files

* add task related questions and reorder the list of task dictionaries

* new tasks, new entries

* eva capabilities configs

* comment out channel related entries for now

* erase old file

* add atmosphere YAML and have model-specific option for the suite

* atmosphere channels

* atmosphere works

* some fixes, ignore certain atmosphere outputs

* add timestamp to cycle_done

* minor changes

* erase line

* minor fix

* fix the call

* gsi nlats

* gsibec coordinates

* Handle task aborts using exceptions (#620)

* Revise error handling

* fixes

* remove unused imports

* revise code tests and fix for git issue

* clean up

* revise change to task_base

* Clean up logger and remove SwellError

* Update src/swell/utilities/filehandler.py

Co-authored-by: Alexey Shiklomanov <alexey.shiklomanov@gmail.com>

* Update src/swell/utilities/filehandler.py

Co-authored-by: Alexey Shiklomanov <alexey.shiklomanov@gmail.com>

* Add more specific exceptions

* Add more specific exceptions

---------

Co-authored-by: Alexey Shiklomanov <alexey.shiklomanov@gmail.com>

* fix leftover wrap arguments (#628)

* remove wrap arguments from logger (#630)

* fix leftover wrap arguments

* clean up last of leftover args

* Add tier1 config for LocalEnsembleDA (#598)

* WIP:  increment-geos_atmosphere.yaml for Eva in lensda

* change:
 - mem: 32
 - window: 3d
 - obs: atms_n20

* Add eva_increment to suites/localensembleda
Restore ens: mem=32, window=3D

* Set background_time_offset("PT3H") to get bkg from 21Z to 3Z

* update res + mem + node to fit ensDA to t1test

* update

* update

* Add config_tier2 in localensembleda

* ...

* add tier1 config for LocalEnsembleDA

* run pycodestyle

* try to make changes to eva  not to lensda

* update

* minimal changes to add eva script for lensda increment

* trick set `max allowable geometry difference:    1.e15` for geometry in bkg files;
this will avoid fv3-jedi geometry code crash.

* turn off linear observer for now

* erase linear observer switch

* no obs space outputs, skip saving obsdiags for now

* bump JEDI build version

* ignore the .venv folder

* no comment zone

* Combine multiple PRs (#629)

* Allow for use of 3d-pressures in interface to observers

* mass requires mid-lev-pressures which are not available in linear fv3-jedi operators

* Revert "trick set `max allowable geometry difference:    1.e15` for geometry in bkg files;"

This reverts commit 8aef9b90d908cb94f9c4cf30d9d6926670bd2c76.

* change gsi_n* defaults

* Handle task aborts using exceptions (#620)

* Revise error handling

* fixes

* remove unused imports

* revise code tests and fix for git issue

* clean up

* revise change to task_base

* Clean up logger and remove SwellError

* Update src/swell/utilities/filehandler.py

Co-authored-by: Alexey Shiklomanov <alexey.shiklomanov@gmail.com>

* Update src/swell/utilities/filehandler.py

Co-authored-by: Alexey Shiklomanov <alexey.shiklomanov@gmail.com>

* Add more specific exceptions

* Add more specific exceptions

---------

Co-authored-by: Alexey Shiklomanov <alexey.shiklomanov@gmail.com>

* fix leftover wrap arguments (#628)

---------

Co-authored-by: Ricardo Todling <ricardo.todling@nasa.gov>
Co-authored-by: Michael Anstett <michael.anstett@nasa.gov>
Co-authored-by: Alexey Shiklomanov <alexey.shiklomanov@gmail.com>

---------

Co-authored-by: dooruk <dorukhan.ardag@nasa.gov>
Co-authored-by: Doruk Ardağ <38666458+Dooruk@users.noreply.github.com>
Co-authored-by: Ricardo Todling <ricardo.todling@nasa.gov>
Co-authored-by: Michael Anstett <michael.anstett@nasa.gov>
Co-authored-by: Alexey Shiklomanov <alexey.shiklomanov@gmail.com>

* Add docs for csh interactive install (#631)

* Initial commit

* Update csh install instructions

* Update installing_swell_csh_interactive.md

* Update installing_swell_csh_interactive.md

* fix order of steps

* Update installing_swell_uv_venv.md

* Feature/mjkim/add tsinc2eva (#632)

* Adding tskin to Eva's default PNG output for increments

* add ts to eva increment plot

---------

Co-authored-by: Doruk Ardağ <38666458+Dooruk@users.noreply.github.com>

---------

Co-authored-by: Cheng Da <111095732+gmao-cda@users.noreply.github.com>
Co-authored-by: Doruk Ardağ <38666458+Dooruk@users.noreply.github.com>
Co-authored-by: rtodling <ricardo.todling@nasa.gov>
Co-authored-by: Yonggang Yu <yonggang.g.yu@gmail.com>
Co-authored-by: gmao-yzhu <Yanqiu.Zhu@nasa.gov>
Co-authored-by: gmao-yzhu <gmao-yzhu@users.noreply.github.com>
Co-authored-by: Alexey Shiklomanov <alexey.shiklomanov@gmail.com>
Co-authored-by: dooruk <dorukhan.ardag@nasa.gov>
Co-authored-by: Min-Jeong Kim <48491996+mjkagnes123@users.noreply.github.com>

mranst committed:
Merge branch 'develop' into feature/iodaconverters

mranst committed:
code test fix

cohen-seth reviewed (commented):
Reviewed all recent changes again. Everything looks good. Everything was still working when I tested it.

Dooruk reviewed (approved):
Looks good 👍🏼

Dooruk committed:
Merge branch 'develop' into feature/iodaconverters

Dooruk committed:
Merge branch 'develop' into feature/iodaconverters
